File: Graphs - DFS/1466. Reorder Routes to Make All Paths Lead to the City Zero/solution.py
class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        graph = defaultdict(list)
        roads = set()
        for u, v in connections:
            graph[u].append(v)
            graph[v].append(u)
            roads.add((u, v))
        visited = [0] * n
        self.res = 0
        def dfs(node):
            visited[node] = 1
            for nex in graph[node]:
                if visited[nex] == 0:
                    if (node, nex) in roads:
                        self.res += 1
                    dfs(nex)
        dfs(0)
        return self.res

File: Graphs - DFS/1466. Reorder Routes to Make All Paths Lead to the City Zero/solution.cs
public class Solution {
    public int MinReorder(int n, int[][] connections) {
        var graph = new Dictionary<int, List<int>>();
        var roads = new HashSet<string>();
        foreach (var connection in connections) {
            if (!graph.ContainsKey(connection[0])) graph[connection[0]] = new List<int>();
            if (!graph.ContainsKey(connection[1])) graph[connection[1]] = new List<int>();
            graph[connection[0]].Add(connection[1]);
            graph[connection[1]].Add(connection[0]);
            roads.Add(connection[0] + "," + connection[1]);
        }
        var visited = new bool[n];
        int res = 0;
        void Dfs(int node) {
            visited[node] = true;
            foreach (var nex in graph[node]) {
                if (!visited[nex]) {
                    if (roads.Contains(node + "," + nex)) res++;
                    Dfs(nex);
                }
            }
        }
        Dfs(0);
        return res;
    }
}